File: UserServer/src/main/java/cc/nobrain/dev/userserver/domain/train/entity/TrainFile.java
package cc.nobrain.dev.userserver.domain.train.entity;

import cc.nobrain.dev.userserver.domain.base.entity.File;
import cc.nobrain.dev.userserver.domain.workspace.entity.Workspace;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.DynamicUpdate;

import java.util.Objects;

@Entity
@DynamicUpdate
@DiscriminatorValue("TRAIN_FILE")
@Getter
@NoArgsConstructor
public class TrainFile extends File {
    @ManyToOne
    @JoinColumn(name = "owner_index")
    protected Workspace ownerIndex;

    @Override
    public <T> void setRelation(T ownerEntity) {
        if (!(ownerEntity instanceof Workspace owner) || Objects.isNull(ownerEntity)) {
            throw new IllegalArgumentException("Invalid owner entity");
        }
        this.ownerIndex = owner;
        owner.getFiles().add(this);
    }
}

File: UserServer/src/main/java/cc/nobrain/dev/userserver/domain/train/entity/TrainFile.kt
package cc.nobrain.dev.userserver.domain.train.entity;

import cc.nobrain.dev.userserver.domain.base.entity.File;
import cc.nobrain.dev.userserver.domain.workspace.entity.Workspace;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.DynamicUpdate;

import java.util.Objects;

@Entity
@DynamicUpdate
@DiscriminatorValue("TRAIN_FILE")
class TrainFile(
    @ManyToOne
    @JoinColumn(name = "ownerIndex", foreignKey = ForeignKey(name = "fk_ownerIndex", value = ConstraintMode.NO_CONSTRAINT))
    var ownerIndex: Workspace? = null,

    @Column(nullable = false, columnDefinition = "varchar(255) default 'none'")
    var label: String = "none",
) : File() {
    override fun <T> setRelation(ownerEntity: T) {
        if (ownerEntity == null || ownerEntity !is Workspace) {
            throw IllegalArgumentException("Invalid owner entity")
        }
        this.ownerIndex = ownerEntity
        if (ownerEntity.files.isNullOrEmpty()) {
            ownerEntity.files = mutableListOf()
        }
        ownerEntity.files.add(this)
    }
}